lib/model/CEventRateModel.cc

```diff
@@ -325,25 +325,26 @@ void CEventRateModel::sample(core_t::TTime startTime,
                                   priorWeights[0]);
                 maths_t::setWinsorisationWeight(winsorisationWeight, priorWeights[0]);
 
+                auto annotationCallback = [&](const std::string& annotation) {
+                    if (this->params().s_AnnotationsEnabled) {
+                        m_CurrentBucketStats.s_Annotations.emplace_back(
+                            time, annotation, gatherer.searchKey().detectorIndex(),
+                            gatherer.searchKey().partitionFieldName(),
+                            gatherer.partitionFieldValue(),
+                            gatherer.searchKey().overFieldName(), EMPTY_STRING,
+                            gatherer.searchKey().byFieldName(), gatherer.personName(pid));
+                    }
+                };
+
                 maths::CModelAddSamplesParams params;
                 params.integer(true)
                     .nonNegative(true)
                     .propagationInterval(deratedInterval)
                     .trendWeights(trendWeights)
-                    .priorWeights(priorWeights);
-                if (this->params().s_AnnotationsEnabled) {
-                    const auto modelAnnotationCallback =
-                        [&](core_t::TTime t, const std::string& annotation) {
-                            m_CurrentBucketStats.s_Annotations.emplace_back(
-                                t, annotation, gatherer.searchKey().detectorIndex(),
-                                gatherer.searchKey().partitionFieldName(),
-                                gatherer.partitionFieldValue(),
-                                gatherer.searchKey().overFieldName(),
-                                EMPTY_STRING, gatherer.searchKey().byFieldName(),
-                                gatherer.personName(pid));
-                        };
-                    params.annotationCallback(modelAnnotationCallback);
-                }
+                    .priorWeights(priorWeights)
+                    .annotationCallback([&](const std::string& annotation) {
```

Comment:
Why is the `annotationCallback` wrapped in a separate lambda here?
Is it because:
> By passing the callback by reference we avoid this.

?

Reply:
You still have to bind it to a `std::function`. You pass by constant reference and this creates a temporary `std::function` object (for the type erasure for the lambda). If the lambda is small, typically less than or equal to 16 bytes, then this object doesn't allocate any memory. Otherwise it allocates memory to hold the lambda. By wrapping the original lambda in another one which captures it by reference the lambda for which we perform type erasure is only 8 bytes: containing only a reference to original lambda as a member rather than having members for the different values which are referenced in the original lambda. For more information see for example [this](https://blog.demofox.org/2015/02/25/avoiding-the-performance-hazzards-of-stdfunction/) discussion.

Reply:
Wow. Thanks for explanation. In moments like this I appreciate the fact I'm (mostly) a Java developer ;)